From 410f068f5147b54b0459c372c714b26bf657f0b4 Mon Sep 17 00:00:00 2001
From: SUraj Sumangala <surajs@qca.qualcomm.com>
Date: Thu, 15 Dec 2011 10:55:15 +0530
Subject: [PATCH 5/8] nl80211: Add NL80211 commands for BT Coex

This patch adds support for new NL80211 commands and attributes for BT Coex.

Signed-off-by: Suraj Sumangala <surajs@qca.qualcomm.com>
---
 include/linux/nl80211.h |  115 +++++++++++++++++++++++++++++++++++++++++++++++
 1 files changed, 115 insertions(+), 0 deletions(-)

diff --git a/include/linux/nl80211.h b/include/linux/nl80211.h
index f9261c2..6ba76ee 100644
--- a/include/linux/nl80211.h
+++ b/include/linux/nl80211.h
@@ -538,6 +538,41 @@
  *	OLBC handling in hostapd. Beacons are reported in %NL80211_CMD_FRAME
  *	messages. Note that per PHY only one application may register.
  *
+ * @NL80211_CMD_BTCOEX_INQ: This command is used to provide WiFi driver the
+ *	Bluetooth inquiry status. The status will be available in flag
+ *	%NL80211_ATTR_BTCOEX_INQ_STATUS. This information can be used to
+ *	manage shared resources when the wireless device is a Bluetooth-Wifi
+ *	coex solution.
+ *
+ * @NL80211_CMD_BTCOEX_SCO: This command is used to give the driver the
+ *	Bluetooth SCO connection status. The SCO status is available in
+ *	%NL80211_ATTR_BTCOEX_SCO_STATUS flag. It also provide
+ *	%NL80211_ATTR_BTCOEX_TYPE_ESCO to specify if the connection is ESCO.
+ *	It also has %NL80211_ATTR_BTCOEX_ESCO_TX_INTERVAL specifing the time
+ *	between consecutive eSCO instance(Invalid for SCO).
+ *	%NL80211_ATTR_BTCOEX_ESCO_TX_PKT_LEN spedicying the the length in
+ *	bytes of the eSCO payload in transmit direction.This feature
+ *	useful for a Bluetooth-Wifi coex solution.
+ *
+ * @NL80211_CMD_BTCOEX_A2DP: This command is used to give the driver the
+ *	Bluetooth A2DP profile connection status. The A2DP profile connection
+ *	status is available in %NL80211_ATTR_BTCOEX_A2DP_STATUS flag.
+ *	This feature is typically used when the wireless device is a
+ *	Bluetooth-Wifi coex solution.
+ *
+ * @NL80211_CMD_BTCOEX_ACL_INFO: This command is used to let the wifi driver
+ *	know information regarding the ACL link. Currently supported
+ *	information includes The role as %NL80211_ATTR_BTCOEX_ACL_ROLE,
+ *	%NL80211_ATTR_BTCOEX_REMOTE_LMP_VER showing the LMP version of the
+ *	remote device. This feature is useful when the wireless device is a
+ *	Bluetooth-Wifi coex solution.
+ *
+ * @NNL80211_CMD_BTCOEX_ANTENNA_CONFIG: This command is used to let the wifi
+ *	driver know information regarding the antenna configuration used
+ *	in case of a BT-coex solution. This information is provide as
+ *	%NL80211_ATTR_BTCOEX_ANTENNA_CONFIG.This feature is useful when
+ *	the wireless device is a Bluetooth-Wifi coex solution.
+ *
  * @NL80211_CMD_MAX: highest used command number
  * @__NL80211_CMD_AFTER_LAST: internal use
  */
@@ -675,6 +710,12 @@ enum nl80211_commands {
 
 	NL80211_CMD_UNEXPECTED_4ADDR_FRAME,
 
+	NL80211_CMD_BTCOEX_INQ,
+	NL80211_CMD_BTCOEX_SCO,
+	NL80211_CMD_BTCOEX_A2DP,
+	NL80211_CMD_BTCOEX_ACL_INFO,
+	NL80211_CMD_BTCOEX_ANTENNA_CONFIG,
+
 	/* add new commands above here */
 
 	/* used to define NL80211_CMD_MAX below */
@@ -1170,6 +1211,50 @@ enum nl80211_commands {
  *	probe-response frame. The DA field in the 802.11 header is zero-ed out,
  *	to be filled by the FW.
  *
+ * @%NL80211_ATTR_BTCOEX_INQ_STATUS: A flag indicating if Bluetooth inquiry
+ *	is in progress. this flag is useful for resource management in a
+ *	bluetooth wifi combo solution.
+ *
+ * @%NL80211_ATTR_BTCOEX_SCO_STATUS: A flag indicating if Bluetooth SCO audio
+ *	connection is active. this flag is useful for resource management in a
+ *	bluetooth wifi combo solution.
+ *
+ * @%NL80211_ATTR_BTCOEX_TYPE_ESCO: A flag indicating if the audio connection
+ *	is of type ESCO. If this attribute is available, the audio connection
+ *	is of type ESCO. this information is useful for resource management
+ *	in a bluetooth wifi combo solution.
+ *
+ * @%NL80211_ATTR_BTCOEX_ESCO_TX_INTERVAL: Provides the time between two
+ *	consecutive eSCO instant, measured in slots.This attribute will be
+ *	available only in case of an eSCO connection.
+ *	this information is useful for resource management in a bluetooth
+ *	wifi combo solution.
+ *
+ * @%NL80211_ATTR_BTCOEX_ESCO_TX_PKT_LEN: Provides the length in bytes of the
+ *	eSCO payload in the receive direction. This attribute will be available
+ *	in case of an eSCO connection.
+ *	this information is useful for resource management in a bluetooth
+ *	wifi combo solution.
+ *
+ * @%NL80211_ATTR_BTCOEX_A2DP_STATUS: A flag indicating the Bluetooth
+ *	A2DP connection status. This flag is useful for resource management
+ *	in a bluetooth wifi combo solution.
+ *
+ * @%NL80211_ATTR_BTCOEX_ACL_ROLE: Indicates if Bluetooth chip's role
+ *	in an ACL connection. See &enum nl80211_btcoex_acl_role for possible
+ *	value. This flag is useful for resource management in a bluetooth
+ *	wifi combo solution.
+ *
+ * @%NL80211_ATTR_BTCOEX_REMOTE_LMP_VER: Indicates the remote device LMP version
+ *	in an ACL connection. See Link manager version parameter in Bluetooth
+ *	assigned numbers for possible value. This values is useful for resource
+ *	management in a bluetooth wifi combo solution.
+ * @%NL80211_ATTR_BTCOEX_ANTENNA_CONFIG: Indicates the Bluetooth wifi chip
+ *	configuration. See &enum nl80211_btcoex_antenna_config for possible
+ *	value. This flag is useful for resource management in a bluetooth
+ *	wifi combo solution.
+ *
+ *
  * @NL80211_ATTR_MAX: highest attribute number currently defined
  * @__NL80211_ATTR_AFTER_LAST: internal use
  */
@@ -1408,6 +1493,16 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_PROBE_RESP,
 
+	NL80211_ATTR_BTCOEX_INQ_STATUS,
+	NL80211_ATTR_BTCOEX_SCO_STATUS,
+	NL80211_ATTR_BTCOEX_TYPE_ESCO,
+	NL80211_ATTR_BTCOEX_ESCO_TX_INTERVAL,
+	NL80211_ATTR_BTCOEX_ESCO_TX_PKT_LEN,
+	NL80211_ATTR_BTCOEX_A2DP_STATUS,
+	NL80211_ATTR_BTCOEX_ACL_ROLE,
+	NL80211_ATTR_BTCOEX_REMOTE_LMP_VER,
+	NL80211_ATTR_BTCOEX_ANTENNA_CONFIG,
+
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
@@ -2761,4 +2856,24 @@ enum nl80211_probe_resp_offload_support_attr {
 	NL80211_PROBE_RESP_OFFLOAD_SUPPORT_80211U =	1<<3,
 };
 
+/*
+ * enum nl80211_btcoex_acl_role - Bluetooth ACL link role
+ * @NL80211_BTCOEX_ACL_ROLE_UNKNOWN: Bluetooth chip role unknown.
+ * @NL80211_BTCOEX_ACL_ROLE_MASTER: Bluetooth chip is in master role.
+ * @NL80211_BTCOEX_ACL_ROLE_SLAVE: Bluetooth chip is in slave role.
+ */
+enum nl80211_btcoex_acl_role {
+	NL80211_BTCOEX_ACL_ROLE_UNKNOWN,
+	NL80211_BTCOEX_ACL_ROLE_MASTER,
+	NL80211_BTCOEX_ACL_ROLE_SLAVE
+};
+/**
+ * enum nl80211_btcoex_antenna_config - Bluetooth WiFi antenna configuration
+ * @NL80211_BTCOEX_ANTENNA_DA: Bluetooth and WiFI chip used dual antenna.
+ * @NL80211_BTCOEX_ANTENNA_SA: Bluetooth and WiFI chip used dual antenna.
+ */
+enum nl80211_btcoex_antenna_config {
+	NL80211_BTCOEX_ANTENNA_DA,
+	NL80211_BTCOEX_ANTENNA_SA,
+};
 #endif /* __LINUX_NL80211_H */
-- 
1.7.7.4

